frame/assets/src/functions.rs

```diff
@@ -354,16 +365,23 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 
 		T::Currency::unreserve(&depositor, deposit);
 
-		if let Remove = Self::dead_account(&who, &mut details, &account.reason, false) {
-			Account::<T, I>::remove(id, &who);
-		} else {
-			debug_assert!(false, "refund did not result in dead account?!");
-			// deposit has been refunded, need to update `Account`
+		debug_assert!(
+			Remove == Self::dead_account(&who, &mut details, &account.reason, false),
+			"refund did not result in dead account?!"
+		);
+
+		if !account.balance.is_zero() && &depositor != who {
+			// Should not fail as a possibility of a reference increment was checked by
+			// `can_increase`.
+			account.reason =
+				Self::new_account(who, &mut details, None).map_err(|_| Error::<T, I>::WouldBurn)?;
 			Account::<T, I>::insert(id, &who, account);
+		} else {
+			Account::<T, I>::remove(id, &who);
+			// Executing a hook here is safe, since it is not in a `mutate`.
+			T::Freezer::died(id, &who);
```

Comment:
This may not be safe in this order. Best to do it strictly at the end, after the `Asset::<T, I>::insert(&id, details);` line.